### PR TITLE
[WIP]: Refactor DETR

### PR DESCRIPTION
**Pull request by Li-Qingyun**
## Motivation

We (me, @jshilong, @LYMDLUT, and @KeiChiTse) are going to refactor DETR-like models, to enhance the usability and readability of our codebase.

### CODE REVIEW

**Comment by Li-Qingyun** [comment]
## Short-term Plan
We plan to implement a base detector class for Transformer-based detector, TransformerDetector, move the transformer-related part in the original head module to the new detector module, and use TransformerEncoder and TransformerDecoder directly in the detector.

File: configs/detr/detr_r50_8x2_150e_coco.py [hunk 1]
```diff
@@ -13,45 +14,34 @@
...
+            self_attn_cfg=dict(embed_dims=256, num_heads=8, dropout=0.1),
+            ffn_cfg=dict(
+                embed_dims=256,
+                feedforward_channels=2048,
+                num_fcs=2,
+                ffn_drop=0.1,
+                act_cfg=dict(type='ReLU', inplace=True)))),
+    decoder_cfg=dict(
+        num_layers=6,
+        layers_cfg=dict(
```
Comment (jshilong) [inline configs/detr/detr_r50_8x2_150e_coco.py:29]: layer_cfg

File: mmdet/models/detectors/detection_transformer.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+from mmdet.core import bbox2result
+from ..builder import DETECTORS, build_backbone, build_head, build_neck
+from ..utils.positional_encoding import SinePositionalEncoding
+from ..utils.transformer import DetrTransformerDecoder, DetrTransformerEncoder
+from .base import BaseDetector
+
+
+@DETECTORS.register_module()
+class TransformerDetector(BaseDetector):
```
Comment (jshilong) [inline mmdet/models/detectors/detection_transformer.py:17]: We  can just define the interface in this class and move all the detail implementation to a specific detector

File: mmdet/models/utils/transformer.py [hunk 3]
```diff
@@ -404,659 +390,230 @@ def inverse_sigmoid(x, eps=1e-5):
...
-
-@TRANSFORMER_LAYER_SEQUENCE.register_module()
-class DetrTransformerEncoder(TransformerLayerSequence):
-    """TransformerEncoder of DETR.
-
-    Args:
-        post_norm_cfg (dict): Config of last normalization layer. Default：
-            `LN`. Only used when `self.pre_norm` is `True`
-    """
+                 layers_cfg=None,
```
Comment (jshilong) [inline mmdet/models/utils/transformer.py:396]: layer_cfg

File: configs/detr/detr_r50_8x2_150e_coco.py (hunk 1 above)
Comment (jshilong) [reply configs/detr/detr_r50_8x2_150e_coco.py:29]: we can discuss whether to support building different layer

File: mmdet/models/utils/transformer.py [hunk 4]
```diff
@@ -404,659 +390,230 @@ def inverse_sigmoid(x, eps=1e-5):
...
+            self_attn_cfg['batch_first'] = batch_first
+        self.batch_first = batch_first  # TODO
+        self.self_attn_cfg = self_attn_cfg  # TODO
+        self.ffn_cfg = ffn_cfg  # TODO
+        self.norm_cfg = norm_cfg  # TODO
+        self._init_layers()
+
+    def _init_layers(self):
+        self.self_attn = MultiheadAttention(**self.self_attn_cfg)
+        self.self_attn.operation_name = 'self_attn'
```
Comment (jshilong) [inline mmdet/models/utils/transformer.py:498]: we can remove this now

File: mmdet/models/utils/transformer.py [hunk 5]
```diff
@@ -404,659 +390,230 @@ def inverse_sigmoid(x, eps=1e-5):
...
-        Args:
-            spatial_shapes (Tensor): The shape of all
-                feature maps, has shape (num_level, 2).
-            valid_ratios (Tensor): The radios of valid
-                points on the feature map, has shape
-                (bs, num_levels, 2)
-            device (obj:`device`): The device where
-                reference_points should be.
+                 self_attn_cfg=dict(
+                     type='MultiheadAttention',
```
Comment (jshilong) [inline mmdet/models/utils/transformer.py:534]: remove the type

File: mmdet/models/utils/transformer.py [hunk 6]
```diff
@@ -404,659 +390,230 @@ def inverse_sigmoid(x, eps=1e-5):
...
+        self.cross_attn_cfg = cross_attn_cfg
+        self.ffn_cfg = ffn_cfg
+        self.norm_cfg = norm_cfg
+        self._init_layers()
+
+    def _init_layers(self):
+        self.self_attn = MultiheadAttention(**self.self_attn_cfg)
+        self.self_attn.operation_name = 'self_attn'
+        self.cross_attn = MultiheadAttention(**self.cross_attn_cfg)
+        self.cross_attn.operation_name = 'cross_attn'
```
Comment (jshilong) [inline mmdet/models/utils/transformer.py:572]: remove this

File: mmdet/models/detectors/detection_transformer.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    def forward_train(self,
+                      img,
+                      img_metas,
+                      gt_bboxes,
+                      gt_labels,
+                      gt_bboxes_ignore=None):
+        super(TransformerDetector, self).forward_train(img, img_metas)
+        x = self.extract_feat(img)
+        assert isinstance(x, tuple) and len(x) == 1  # TODO: delete this
+        x, mask, pos_embed = self.forward_pretransformer(x[0], img_metas)
```
Comment (jshilong) [inline mmdet/models/detectors/detection_transformer.py:105]: out = self.forward_pretransformer
outs_dec, _ = self.forward_transformer(*out，
                                               pos_embed)

File: mmdet/models/utils/transformer.py [hunk 8]
```diff
@@ -404,659 +390,229 @@ def inverse_sigmoid(x, eps=1e-5):
...
-            query=feat_flatten,
-            key=None,
-            value=None,
-            query_pos=lvl_pos_embed_flatten,
-            query_key_padding_mask=mask_flatten,
-            spatial_shapes=spatial_shapes,
-            reference_points=reference_points,
-            level_start_index=level_start_index,
-            valid_ratios=valid_ratios,
+        query = self.self_attn(
```
Comment (chhluo) [inline mmdet/models/utils/transformer.py:592]: In Mask2Former, operation_order=('cross_attn', 'norm', 'self_attn', 'norm',                                  'ffn', 'norm'), see https://github.com/open-mmlab/mmdetection/blob/f8bbba2fe777dd882a8da20625be858a98436703/configs/mask2former/mask2former_r50_lsj_8x2_50e_coco-panoptic.py#L88

Comment (Li-Qingyun) [reply mmdet/models/utils/transformer.py:592]: Hi~ Thanks for your review ! 
The decoder was recommanded to custom decoderlayer module for each transformer detector.
And we were going to move heavy transformer modules to detector.

We have recommit a PR #8763 on dev-3.x, and this PR will be closed.
Additionally, I can assist with future refactor work of the Mask2Former.

For Mask2Former, the refactor decoderlayer can be implemented like:
```Python
class Mask2FormerDecoderLayer(DetrTransformerDecoderLayer):
    # I'm not sure to inherit DetrTransformerDecoderLayer or DeformableDetrTransformerDecoderLayer temporarily.
    def __init__(self, *args, **kwargs):
        super(Mask2FormerDecoderLayer, self).__init__(*args, **kwargs)

    def forward(self,
                query,
                key=None,
                value=None,
                query_pos=None,
                key_pos=None,
                self_attn_masks=None,
                cross_attn_masks=None,
                query_key_padding_mask=None,
                key_padding_mask=None,
                **kwargs):
        query = self.cross_attn(
            query=query,
            key=key,
            value=value,
            query_pos=query_pos,
            key_pos=key_pos,
            attn_mask=cross_attn_masks,
            key_padding_mask=key_padding_mask,
            **kwargs)
        query = self.norms[0](query)
        query = self.self_attn(
            query=query,
            key=query,
            value=query,
            query_pos=query_pos,
            key_pos=query_pos,
            attn_mask=self_attn_masks,
            key_padding_mask=query_key_padding_mask,
            **kwargs)
        query = self.norms[1](query)
        query = self.ffn(query)
        query = self.norms[2](query)

        return query
```

File: mmdet/models/detectors/detection_transformer.py [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+
+    # def _load_from_state_dict  # TODO !
+
+    def extract_feat(self, img):
+        x = self.backbone(img)
+        if self.with_neck:
+            x = self.neck(x)
+        return x
+
+    def forward_train(self,
```
[the remark this answers is not in the document]
Comment (Li-Qingyun) [reply mmdet/models/detectors/detection_transformer.py:96]: Hi~ Thanks for your review !
This suggestion is very important. We will discuss it at our next meeting.

**Comment by Li-Qingyun** [comment]
We decided to PR refactored DETR-related modules on dev-3.x. Hence, this PR may not be updated, but reviews and comments about this PR will continue to be answered. The new PR is #8763.

File: mmdet/models/utils/transformer.py (hunk 8 above)
Comment (chhluo) [reply mmdet/models/utils/transformer.py:592]: Great!

**Comment by jshilong** [comment]
To [[WIP]: Refactor DETR and Deformable DETR on 3.x](https://github.com/open-mmlab/mmdetection/pull/8763#top) #8763

**Pull request closed without merging**